### PR TITLE
Add benchmark

### LINKED ISSUE #8: Figure out how to add [bench] benchmarks.
I tried doing this but it the test library is not stable.

### PR DESCRIPTION
**Pull request by iamonuwa**
<del>Fixes</del><ins>work towards</ins> #8

### CODE REVIEW

**Comment by iamonuwa** [comment]
@mbebenita @yurydelendik please review

File: benches/benchmark.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
+pub fn read_file_data(path: &PathBuf) -> Vec<u8> {
+    println!("Parsing {:?}", path);
```
Comment (yurydelendik) [inline benches/benchmark.rs]: can you remove this line

**Review by yurydelendik — COMMENTED** [review]
When I add sleep call to the read_file_data(), I see changes in the benchmark performance run -- any I/O shall not affect test results.

File: benches/benchmark.rs (hunk 1 above)
Comment (iamonuwa) [reply benches/benchmark.rs]: Okay

**Comment by iamonuwa** [comment]
> When I add sleep call to the read_file_data(), I see changes in the benchmark performance run -- any I/O shall not affect test results.

Can you elaborate on this please?

**Comment by yurydelendik** [comment]
> Can you elaborate on this please?

Benchmark test need to measure performance of the Parser/ValidatingParser. Looks like the solution is measuring the performance of the read_file_data function as well, which is not part of the Parser, and definitely skews the results. As I mentioned in [the comment](https://github.com/yurydelendik/wasmparser.rs/issues/8#issuecomment-471999627), to avoid that you may need to load/cache all wasm data prior running the testing.

**Comment by iamonuwa** [comment]
> > Can you elaborate on this please?
> 
> Benchmark test need to measure performance of the Parser/ValidatingParser. Looks like the solution is measuring the performance of the read_file_data function as well, which is not part of the Parser, and definitely skews the results. As I mentioned in [the comment](https://github.com/yurydelendik/wasmparser.rs/issues/8#issuecomment-471999627), to avoid that you may need to load/cache all wasm data prior running the testing.

Any preferred option for in-memory cache?

**Comment by yurydelendik** [comment]
> Any preferred option for in-memory cache?

`std::vec::Vec<>` shall work fine.

**Comment by iamonuwa** [comment]
Alright, I'll work with this now

**Comment by iamonuwa** [comment]
@yurydelendik If I get you correctly, the benchmark should run for each of the tests? Right

**Comment by yurydelendik** [comment]
>  If I get you correctly, the benchmark should run for each of the tests? Right

I don't understand the question. But what I was saying that performance of read_file_data (which has I/O operations) shall not influence results of the benchmark test results. #8 is not clear on that, but bench shall test performances of the Parser (and ValidatingParser), and we want to have most of its functionality covered during test run. That said, the enough tests shall be executed during bench to get performance coverage of most of the code base. (The performances of reading files from disk is irrelevant)

**Comment by iamonuwa** [comment]
Do you have a link or guide that can explain how to remove the additional functions? This is how I normally write benchmarks

**Comment by iamonuwa** [comment]
This is how the output looks like.
image

I had to move the data outside the bench_function

**Comment by iamonuwa** [comment]
Can you confirm?

**Comment by yurydelendik** [comment]
Tested again: still a problem

```
Yurys-Workspace:wasmparser yury$ cargo bench
    Finished release [optimized] target(s) in 0.18s
     Running target/release/deps/wasmparser-b27d87a6221dce03

running 7 tests
test tests::simple_tests::default_read ... ignored
test tests::simple_tests::default_read_with_input ... ignored
test tests::simple_tests::it_works ... ignored
test tests::simple_tests::skipping ... ignored
test tests::simple_tests::validator_fails ... ignored
test tests::simple_tests::validator_not_fails ... ignored
test validator::test_validate ... ignored

test result: ok. 0 passed; 0 failed; 7 ignored; 0 measured; 0 filtered out

     Running target/release/deps/benchmark-97097cd7f0076b61
Gnuplot not found, disabling plotting
it works benchmark      time:   [115.85 ns 118.91 ns 123.55 ns]                               
Found 8 outliers among 100 measurements (8.00%)
  8 (8.00%) high severe

validator no fails benchmark                                                                            
                        time:   [251.47 ns 256.82 ns 263.57 ns]
Found 9 outliers among 100 measurements (9.00%)
  2 (2.00%) high mild
  7 (7.00%) high severe

Gnuplot not found, disabling plotting
Yurys-Workspace:wasmparser yury$ vi benches/benchmark.rs 
Yurys-Workspace:wasmparser yury$ git diff
diff --git a/benches/benchmark.rs b/benches/benchmark.rs
index b1d600c..39282d0 100644
--- a/benches/benchmark.rs
+++ b/benches/benchmark.rs
@@ -1,6 +1,7 @@
 pub fn read_file_data(path: &PathBuf) -> Vec<u8> {
     let mut data = Vec::new();
     let mut f = File::open(path).ok().unwrap();
+::std::thread::sleep(::std::time::Duration::from_micros(50));
     f.read_to_end(&mut data).unwrap();
     data
 }
Yurys-Workspace:wasmparser yury$ cargo bench
   Compiling wasmparser v0.29.2 (/Users/yury/Work/junk/wasmparser.rs/wasmparser)
    Finished release [optimized] target(s) in 1.02s
     Running target/release/deps/wasmparser-b27d87a6221dce03

running 7 tests
test tests::simple_tests::default_read ... ignored
test tests::simple_tests::default_read_with_input ... ignored
test tests::simple_tests::it_works ... ignored
test tests::simple_tests::skipping ... ignored
test tests::simple_tests::validator_fails ... ignored
test tests::simple_tests::validator_not_fails ... ignored
test validator::test_validate ... ignored

test result: ok. 0 passed; 0 failed; 7 ignored; 0 measured; 0 filtered out

     Running target/release/deps/benchmark-97097cd7f0076b61
Gnuplot not found, disabling plotting
it works benchmark      time:   [232.70 ns 260.96 ns 290.36 ns]                               
                        change: [+124.96% +143.21% +160.95%] (p = 0.00 < 0.05)
                        Performance has regressed.

validator no fails benchmark                                                                            
                        time:   [308.30 ns 347.51 ns 393.16 ns]
                        change: [+22.541% +31.809% +43.493%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 12 outliers among 100 measurements (12.00%)
  2 (2.00%) high mild
  10 (10.00%) high severe

Gnuplot not found, disabling plotting
```

Notice the performance "regressed" because of in slow down in reading of the file from disk.

**Comment by yurydelendik** [comment]
> I had to move the data outside the bench_function

Did you commit this code?

**Comment by iamonuwa** [comment]
Apologies, pushing new update now

File: benches/benchmark.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+use std::path::PathBuf;
+
+fn it_works_benchmark(c: &mut Criterion) {
+    for entry in read_dir("tests").unwrap() {
+        let dir = entry.unwrap();
+        if !dir.file_type().unwrap().is_file() {
+            continue;
+        }
+        let data = read_file_data(&dir.path());
+        c.bench_function("it works benchmark", move |b| {
```
Comment (yurydelendik) [inline benches/benchmark.rs]: (here and below) move bench_function call out of this loop, store loaded file in the vector, add second loop in the closure and iterate on loaded files.

**Review by yurydelendik — CHANGES_REQUESTED** [review]

File: benches/benchmark.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+use criterion::Criterion;
+use wasmparser::{OperatorValidatorConfig, Parser, ValidatingParser, ValidatingParserConfig};
+
+use std::fs::{read_dir, File};
+use std::io::Read;
+use std::path::PathBuf;
+
+fn it_works_benchmark(c: &mut Criterion) {
+    let mut data: Vec<u8> = vec![];
```
Comment (yurydelendik) [inline benches/benchmark.rs]: Use `Vec<Vec<u8>>` here

Comment (iamonuwa) [reply benches/benchmark.rs]: Using this causes failure on the bench

image

Comment (yurydelendik) [reply benches/benchmark.rs]: See https://github.com/yurydelendik/wasmparser.rs/pull/95#discussion_r265293414

> ... store loaded file in the vector, add second loop in the closure and iterate on loaded files.

**Comment by iamonuwa** [comment]
@yurydelendik @mbebenita  please review

File: benches/benchmark.rs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+fn it_works_benchmark(c: &mut Criterion) {
+    let mut data: Vec<Vec<u8>> = vec![vec![]];
+    for entry in read_dir("tests").unwrap() {
+        let dir = entry.unwrap();
+        if !dir.file_type().unwrap().is_file() {
+            continue;
+        }
+        data.push(read_file_data(&dir.path()));
+    }
+    for d in data {
```
Comment (yurydelendik) [inline benches/benchmark.rs]: As pointed in https://github.com/yurydelendik/wasmparser.rs/pull/95/files#r265293414
> add second loop in the closure and iterate on loaded files.

Could you move this loop *inside* the bench_function ?

Comment (iamonuwa) [reply benches/benchmark.rs]: Yeah sure!

**Comment by iamonuwa** [comment]
@yurydelendik please review

**Review by yurydelendik — APPROVED** [review]

**Comment by yurydelendik** [comment]
Thank you for the patch

**Comment by iamonuwa** [comment]
Thank you for supporting!

**Pull request merged**